File: Python/treehopper/api/HardwareUart.py
import logging
from typing import List, TYPE_CHECKING
if TYPE_CHECKING:
    from treehopper.api import TreehopperUsb
from treehopper.api import DeviceCommands
from treehopper.api.IOneWire import IOneWire
from treehopper.utils.error import error


class UartMode:
    Uart = 0
    """The UART is in standard UART mode."""

    OneWire = 1
    """The UART is in OneWire mode"""

class UartConfig:
    Disabled, Standard, OneWire = range(3)


class UartCommand:
    Transmit, Receive, OneWireReset, OneWireScan = range(4)


class HardwareUart(IOneWire):
    """
    Hardware UART module
    """

    def __init__(self, board: 'TreehopperUsb'):
        """
        Construct a new Hardware UART connected to the specified board.
        @param board: the board the hardware UART is attached to

        This constructor should not be called by the user; you can obtain a reference to the hardware UART
        on your board by accessing the board's "uart" property. See the documentation for TreehopperUsb for more info.
        """
        self._board = board
        self._baud = 9600
        self._enabled = False
        self._mode = UartMode.Uart
        self._use_open_drain_tx = False
        self._logging = logging.getLogger(__name__)

    def start_one_wire(self):
        """
        Place the UART into One-Wire Mode and enable the peripheral.
        """
        self.mode = UartMode.OneWire
        self.enabled = True

    def one_wire_reset_and_match_address(self, address: int):
        """
        Reset the bus and address the device specified
        @param address: the device to address
        """
        self.one_wire_reset()
        data = [0x55] + list(address.to_bytes(8, 'big'))
        self.send(data)

    def one_wire_search(self) -> List[int]:
        """
        Search the 1-Wire Bus for devices.
        @return: a list of addresses of devices attached to the bus.
        """
        self.mode = UartMode.OneWire
        self.enabled = True

        ret_val = []  # list[int]

        data = [DeviceCommands.UartTransaction, UartCommand.OneWireScan]
        with self._board._comms_lock:
            self._board._send_peripheral_config_packet(data)
            while True:
                received_data = self._board._receive_comms_response_packet(9)
                if received_data[0] == 0xff:
                    break

                ret_val.append(int.from_bytes(received_data[1:], 'big'))
        return ret_val

    def one_wire_reset(self) -> bool:
        """
        Reset the 1-Wire bus.
        @return: whether a device was found on the bus or not.
        """
        self.mode = UartMode.OneWire
        self.enabled = True
        ret_val = False
        data = [DeviceCommands.UartTransaction, UartCommand.OneWireReset]
        with self._board._comms_lock:
            self._board._send_peripheral_config_packet(data)
            received_data = self._board._receive_comms_response_packet(1)
            ret_val = received_data[0] > 0

        return ret_val

    def receive(self, one_wire_num_bytes=0) -> List[int]:
        """
        Receive characters from the UART
        @param one_wire_num_bytes: in OneWire mode, specifies the number of characters to read. Ignored in standard
        UART mode.
        @return: the received characters.

        Calling this function will return the current UART receive FIFO and clear it. The FIFO supports a maximum of 32
        characters.
        """
        ret_val = []  # List[int]
        if self._mode == UartMode.Uart:
            if one_wire_num_bytes != 0:
                self._logging.info("Since the UART is not in One-Wire Mode, the oneWireNumBytes parameter is ignored")

            data = [DeviceCommands.UartTransaction, UartCommand.Receive]

            with self._board._comms_lock:
                self._board._send_peripheral_config_packet(data)
                received_data = self._board._receive_comms_response_packet(33)
                len = received_data[32]
                ret_val = received_data[0:len]

        else:
            if one_wire_num_bytes == 0:
                error(logging, "You must specify the number of bytes to receive in One-Wire Mode", False)
                return ret_val

            data = [DeviceCommands.UartTransaction, UartCommand.Receive, one_wire_num_bytes]

            with self._board._comms_lock:
                self._board._send_peripheral_config_packet(data)
                received_data = self._board._receive_comms_response_packet(33)
                len = received_data[32]
                ret_val = received_data[0:len]
        return ret_val

    def send(self, data_to_send: List[int]):
        """
        Send data out of the UART
        @param data_to_send: the data to send
        """
        if len(data_to_send) > 63:
            error(logging, "The maximum UART length for one transaction is 63 bytes", True)
        data = [DeviceCommands.UartTransaction, UartCommand.Transmit, len(data_to_send)] + data_to_send
        with self._board._comms_lock:
            self._board._send_peripheral_config_packet(data)
            self._board._receive_comms_response_packet(1)


    @property
    def mode(self):
        """[Property] Gets or sets the mode of the device."""
        return self._mode

    @mode.setter
    def mode(self, value):
        """[Property] Gets or sets the mode of the device."""
        if self._mode == value:
            return

        self._mode = value
        self._update_config()

    @property
    def enabled(self):
        """[Property] Gets or sets whether the device is enabled."""
        return self._enabled

    @enabled.setter
    def enabled(self, value):
        """[Property] Gets or sets whether the device is enabled."""
        if self._enabled == value:
            return

        self._enabled = value
        self._update_config()

    @property
    def baud(self):
        """[Property] Gets or sets the baud of the UART."""
        return self._baud

    @baud.setter
    def baud(self, value):
        """[Property] Gets or sets the baud of the UART."""
        if self._baud == value:
            self._baud = value

        self._update_config()

    def _update_config(self):
        if not self._enabled:
            self._board._send_peripheral_config_packet([DeviceCommands.UartConfig, UartConfig.Disabled])
        elif self._mode == UartMode.Uart:
            timer_val = 0
            use_prescaler = False

            timer_val_prescaler = round(256.0 - 2000000.0 / self._baud)
            timer_val_no_prescaler = round(256.0 - 24000000.0 / self._baud)

            prescaler_out_of_bounds = (timer_val_prescaler > 255) or timer_val_prescaler < 0
            no_prescaler_out_of_bounds = (timer_val_no_prescaler > 255) or (timer_val_no_prescaler < 0)

            prescaler_error = round(self._baud - 2000000 / (256 - timer_val_prescaler))
            no_prescaler_error = round(self._baud - 24000000 / (256 - timer_val_no_prescaler))

            if prescaler_out_of_bounds and no_prescaler_out_of_bounds:
                error(self._logging, "The specified baud rate is out of bounds.", False)

            if prescaler_out_of_bounds:
                use_prescaler = False
                timer_val = timer_val_no_prescaler
            elif no_prescaler_out_of_bounds:
                use_prescaler = True
                timer_val = timer_val_prescaler
            elif prescaler_error > no_prescaler_error:
                use_prescaler = False
                timer_val = timer_val_no_prescaler
            else:
                use_prescaler = True
                timer_val = timer_val_no_prescaler
            
            data = [DeviceCommands.UartConfig, UartConfig.Standard, round(timer_val), use_prescaler, self._use_open_drain_tx]
            self._board._send_peripheral_config_packet(data)
        else:
            data =[DeviceCommands.UartConfig, UartConfig.OneWire]
            self._board._send_peripheral_config_packet(data)

    def __str__(self):
        """
        Get a string representation of the UART's state
        :return: a string representation of the state
        """
        if self._enabled:
            if self._mode == UartMode.Uart:
                return "Uart, running at {} baud".format(self._baud)
            else:
                return "1-Wire"
        else:
            return "Disabled"

File: C++/API/src/HardwareUart.cpp
#include "HardwareUart.h"
#include "TreehopperUsb.h"

void Treehopper::HardwareUart::startOneWire() {
    _mode = UartMode::OneWire;
    _enabled = true;
    updateConfig();
}

void Treehopper::HardwareUart::oneWireResetAndMatchAddress(uint64_t address) {
    mode(UartMode::OneWire);
    enabled(true);
    oneWireReset();
    auto data = Utility::getBytes(address, 8);
    data.insert(data.begin(), 0x55); // MATCH ROM
    send(data);
}

std::vector<uint64_t> Treehopper::HardwareUart::oneWireSearch() {

    mode(UartMode::OneWire);
    enabled(true);

    std::vector<uint64_t> retVal;

    uint8_t data[] = {(uint8_t)TreehopperUsb::DeviceCommands::UartTransaction,
                      (uint8_t)UartCommand::OneWireScan};

    {
        _device.sendPeripheralConfigPacket(data, sizeof(data));

        while (true) {
            uint8_t receivedData[9];
            _device.receivePeripheralConfigPacket(receivedData, 9);
            if (receivedData[0] == 0xff)
                break;

            std::vector<uint8_t> receivedWord(&receivedData[1], &receivedData[9]);
            retVal.push_back(Utility::getValue(receivedWord, true));
        }
    }
    return retVal;
}

bool Treehopper::HardwareUart::oneWireReset() {
    mode(UartMode::OneWire);
    enabled(true);
    bool retVal;
    uint8_t data[] = {
            (uint8_t) TreehopperUsb::DeviceCommands::UartTransaction,
            (uint8_t) UartCommand ::OneWireReset
    };

    {
        _device.sendPeripheralConfigPacket(data, 2);
        uint8_t receivedData;
        _device.receivePeripheralConfigPacket(&receivedData, 1);
        retVal = receivedData > 0;
    }

    return retVal;
}

std::vector<uint8_t> Treehopper::HardwareUart::receive(int oneWireNumBytes) {

    uint8_t receivedData[33];
    int len;

    if (_mode == UartMode::Uart)
    {
        if(oneWireNumBytes != 0)
            Utility::error("Since the UART is not in One-Wire Mode, the oneWireNumBytes parameter is ignored");
        uint8_t data[] = {
                (uint8_t)TreehopperUsb::DeviceCommands::UartTransaction,
                (uint8_t)UartCommand::Receive
        };
        {
            _device.sendPeripheralConfigPacket(data, 2);
            _device.receivePeripheralConfigPacket(receivedData, 33);
            len = receivedData[32];
        }
    }
    else
    {
        if (oneWireNumBytes == 0)
            Utility::error("You must specify the number of bytes to receive in One-Wire Mode");
        uint8_t data[] = {
                (uint8_t) TreehopperUsb::DeviceCommands::UartTransaction,
                (uint8_t)UartCommand ::Receive,
                (uint8_t)oneWireNumBytes
        };
        {
            _device.sendPeripheralConfigPacket(data, 3);
            _device.receivePeripheralConfigPacket(receivedData, 33);
            len = receivedData[32];
        }
    }

    return std::vector<uint8_t>(receivedData, (receivedData + len));
}

void Treehopper::HardwareUart::send(std::vector<uint8_t> dataToSend) {
    if (dataToSend.size() > 63)
        Utility::error("The maximum UART length for one transaction is 63 bytes", true);

    uint8_t data[] = {
        (uint8_t)TreehopperUsb::DeviceCommands::UartTransaction,
        (uint8_t)UartCommand::Transmit,
        dataToSend.size()
    };

    dataToSend.insert(dataToSend.begin(), (uint8_t)TreehopperUsb::DeviceCommands::UartTransaction);
    dataToSend.insert(dataToSend.begin()+1, (uint8_t)UartCommand::Transmit);
    dataToSend.insert(dataToSend.begin()+2, dataToSend.size());

    uint8_t receiveData;
    {
        _device.sendPeripheralConfigPacket(dataToSend.data(), dataToSend.size());
        _device.receivePeripheralConfigPacket(&receiveData, 1);
    }
}

void Treehopper::HardwareUart::send(uint8_t dataToSend) {
    std::vector<uint8_t> data = {dataToSend};
    return send(data);
}

void Treehopper::HardwareUart::startUart() {
    _mode = UartMode::Uart;
    _enabled = true;
    updateConfig();
}

std::vector<uint8_t> Treehopper::HardwareUart::receive() {
    std::vector<uint8_t> receivedData;
    return receivedData;
}

void Treehopper::HardwareUart::baud(int baud) {
    if(_baud == baud)
        return;

    _baud = baud;
    updateConfig();
}

int Treehopper::HardwareUart::baud() {
    return _baud;
}

void Treehopper::HardwareUart::mode(Treehopper::UartMode mode) {
    if(_mode == mode)
        return;

    _mode = mode;
    updateConfig();
}

Treehopper::UartMode Treehopper::HardwareUart::mode() {
    return _mode;
}

void Treehopper::HardwareUart::enabled(bool enabled) {
    if(_enabled == enabled)
        return;

    _enabled = enabled;
    updateConfig();
}

bool Treehopper::HardwareUart::enabled() {
    return _enabled;
}

void Treehopper::HardwareUart::updateConfig() {
    if (!_enabled)
    {
        uint8_t dataToSend[2] =
                {(uint8_t)TreehopperUsb::DeviceCommands::UartConfig,
                 (uint8_t)UartConfig::Disabled};

        _device.sendPeripheralConfigPacket(dataToSend, 2);
    }
    else if (_mode == UartMode::Uart)
    {
        uint8_t timerVal;
        bool usePrescaler;

        // calculate baud with and without prescaler
        auto timerValPrescaler = (int) round(256.0 - 2000000.0 / _baud);
        auto timerValNoPrescaler = (int) round(256.0 - 24000000.0 / _baud);

        auto prescalerOutOfBounds = timerValPrescaler > 255 || timerValPrescaler < 0;
        auto noPrescalerOutOfBounds = timerValNoPrescaler > 255 || timerValNoPrescaler < 0;

        // calculate error
        double prescalerError = abs(_baud - 2000000 / (256 - timerValPrescaler));
        double noPrescalerError = abs(_baud - 24000000 / (256 - timerValNoPrescaler));

        if (prescalerOutOfBounds && noPrescalerOutOfBounds)
            Utility::error("The specified baud rate is out of bounds.", true);
        if (prescalerOutOfBounds)
        {
            usePrescaler = false;
            timerVal = (uint8_t) timerValNoPrescaler;
        }
        else if (noPrescalerOutOfBounds)
        {
            usePrescaler = true;
            timerVal = (uint8_t) timerValPrescaler;
        }
        else if (prescalerError > noPrescalerError)
        {
            usePrescaler = false;
            timerVal = (uint8_t) timerValNoPrescaler;
        }
        else
        {
            usePrescaler = true;
            timerVal = (uint8_t) timerValPrescaler;
        }

        uint8_t data[5];
        data[0] = (uint8_t) TreehopperUsb::DeviceCommands::UartConfig;
        data[1] = (uint8_t) UartConfig::Standard;
        data[2] = timerVal;
        data[3] = (uint8_t) (usePrescaler ? 0x01 : 0x00);
        data[4] = (uint8_t) (_useOpenDrainTx ? 0x01 : 0x00);
        _device.sendPeripheralConfigPacket(data, sizeof(data));
    }
    else
    {
        uint8_t data[2];
        data[0] = (uint8_t) TreehopperUsb::DeviceCommands::UartConfig;
        data[1] = (uint8_t) UartConfig::OneWire;
        _device.sendPeripheralConfigPacket(data, sizeof(data));
    }
}
